File: src/routes/catalog.ts
import express, { Response, NextFunction } from 'express';
import { body, validationResult } from 'express-validator';
import { authenticate, type AuthRequest } from '../middleware/auth.js';
import type { BlockDefinition } from '../data/blockCatalog.js';
import { BLOCK_CATALOG } from '../data/blockCatalog.js';

const router = express.Router();

router.use(authenticate);

let runtimeCatalog: BlockDefinition[] = [...BLOCK_CATALOG];

router.get('/blocks', (_req, res: Response) => {
  res.json({ blocks: runtimeCatalog });
});

router.post(
  '/blocks',
  [
    body('id').trim().notEmpty(),
    body('name').trim().notEmpty(),
    body('category').trim().isIn(['kitchen', 'bathroom', 'furniture', 'electrical', 'plumbing']),
    body('width').isNumeric(),
    body('height').isNumeric(),
  ],
  (req: AuthRequest, res: Response, next: NextFunction) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({ errors: errors.array() });
      }
      const block: BlockDefinition = {
        id: req.body.id,
        name: req.body.name,
        type: 'furniture',
        category: req.body.category,
        manufacturer: req.body.manufacturer,
        sku: req.body.sku,
        tags: req.body.tags,
        moduleClass: req.body.moduleClass,
        width: Number(req.body.width),
        height: Number(req.body.height),
        depth: req.body.depth ? Number(req.body.depth) : undefined,
        description: req.body.description,
        planSymbols: req.body.planSymbols || [],
      };

      runtimeCatalog = runtimeCatalog.filter((existing) => existing.id !== block.id).concat(block);

      res.status(201).json({ block });
    } catch (error) {
      next(error);
    }
  },
);

export default router;

File: app/routers/catalog.py
# app/routers/catalog.py
from fastapi import APIRouter, Depends
from pydantic import BaseModel
from typing import Optional, Any, List
from app.middleware.auth import get_current_user

router = APIRouter()

class BlockDefinition(BaseModel):
    id: str
    name: str
    type: str = "furniture"
    category: str
    manufacturer: Optional[str] = None
    sku: Optional[str] = None
    tags: Optional[List[str]] = None
    moduleClass: Optional[str] = None
    width: float
    height: float
    depth: Optional[float] = None
    description: Optional[str] = None
    planSymbols: Optional[List[Any]] = None

# In-memory catalog (can be replaced with database)
runtime_catalog: List[BlockDefinition] = []

@router.get("/blocks")
async def get_blocks(current_user: dict = Depends(get_current_user)):
    """Get all catalog blocks"""
    return {"blocks": runtime_catalog}

@router.post("/blocks")
async def create_block(block: BlockDefinition, current_user: dict = Depends(get_current_user)):
    """Create or update a catalog block"""
    global runtime_catalog
    
    # Remove existing block with same id
    runtime_catalog = [b for b in runtime_catalog if b.id != block.id]
    
    # Add new block
    runtime_catalog.append(block)
    
    return {"block": block}

